File: seminar-08/B/python/B.py
import sys

sys.setrecursionlimit(10**6)


# DFS ищет точки сочленения (cutpoints)
def dfs(v, parent, adj, visited, tin, low, timer_ref, cutpoints):
    visited[v] = True
    tin[v] = low[v] = timer_ref[0]
    timer_ref[0] += 1

    children = 0  # число детей v в DFS-дереве

    for to in adj[v]:
        if to == parent:
            continue

        if visited[to]:
            # обратное ребро
            low[v] = min(low[v], tin[to])
        else:
            dfs(to, v, adj, visited, tin, low, timer_ref, cutpoints)
            children += 1

            low[v] = min(low[v], low[to])

            # v — НЕ корень, и поддерево to не может уйти выше v
            if parent != -1 and low[to] >= tin[v]:
                cutpoints.add(v)

    # отдельное правило для корня DFS
    if parent == -1 and children >= 2:
        cutpoints.add(v)


input = sys.stdin.readline
n, m = map(int, input().split())

adj = [[] for _ in range(n)]
for _ in range(m):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    adj[u].append(v)
    adj[v].append(u)

visited = [False] * n
tin = [-1] * n
low = [-1] * n
timer_ref = [0]

cutpoints = set()

# граф может быть несвязным
for v in range(n):
    if not visited[v]:
        dfs(v, -1, adj, visited, tin, low, timer_ref, cutpoints)

ans = sorted(x + 1 for x in cutpoints)
print(len(ans))
if ans:
    print(*ans)

File: seminar-08/B/java/B.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class B {
    static List<Integer>[] adj;
    static boolean[] visited;
    static int[] tin, low;
    static int timer = 0;
    static Set<Integer> cutpoints = new HashSet<>();

    static void dfs(int v, int parent) {
        visited[v] = true;
        tin[v] = low[v] = timer++;
        int children = 0;

        for (int to : adj[v]) {
            if (to == parent) continue;

            if (visited[to]) {
                low[v] = Math.min(low[v], tin[to]);
            } else {
                dfs(to, v);
                children++;

                low[v] = Math.min(low[v], low[to]);

                if (parent != -1 && low[to] >= tin[v]) {
                    cutpoints.add(v);
                }
            }
        }

        if (parent == -1 && children >= 2) {
            cutpoints.add(v);
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] first = br.readLine().split(" ");
        int n = Integer.parseInt(first[0]);
        int m = Integer.parseInt(first[1]);

        adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            String[] parts = br.readLine().split(" ");
            int u = Integer.parseInt(parts[0]) - 1;
            int v = Integer.parseInt(parts[1]) - 1;

            adj[u].add(v);
            adj[v].add(u);
        }

        visited = new boolean[n];
        tin = new int[n];
        low = new int[n];

        Arrays.fill(tin, -1);
        Arrays.fill(low, -1);

        for (int v = 0; v < n; v++) {
            if (!visited[v]) {
                dfs(v, -1);
            }
        }

        List<Integer> ans = new ArrayList<>();
        for (int v : cutpoints) {
            ans.add(v + 1);
        }

        Collections.sort(ans);

        StringBuilder out = new StringBuilder();
        out.append(ans.size()).append('\n');
        if (!ans.isEmpty()) {
            for (int i = 0; i < ans.size(); i++) {
                if (i > 0) out.append(' ');
                out.append(ans.get(i));
            }
            out.append('\n');
        }

        System.out.print(out);
    }
}